### PR TITLE
ci: test build across OS/Node/C++ matrix

### PR DESCRIPTION
**Pull request by jordan-gillard**
# Summary

This PR adds CI to the project via GitHub Actions. It:
* Installs dependencies to exactly match the versions in `package-lock.json` via `npm ci`. This builds `gl`.
* Builds the project via `npm run build`.
* Runs `npm ci` & `npm run build` for Node v18 and Node v20 on the latest versions of Windows and MacOS.
* Runs `npm ci` & `npm run build` for Node v18 and g++ compilers version [9, 13] on Ubuntu.

During this process I confirmed that Node 20 is currently incompatible with the `gl` package dependency on Ubuntu, and that g++ version 8 and older are totally incompatible in both Node 18 and 20 on Ubuntu. See https://github.com/root-project/jsroot/issues/264#issuecomment-1732455658

Closes: #264 

Note: You can see what the CI pipeline looks like on my test PR here: https://github.com/jordan-gillard/jsroot/pull/1

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "jsroot",
   "version": "7.5.0",
   "engines": {
-    "node": ">=0.18.0"
```
Comment (jordan-gillard) [inline package.json:5]: This is fix because `0.18.0` is actually Node's major version 0, and minor version 18 😅

File: package.json [hunk 2]
```diff
@@ -68,7 +68,7 @@
     "eslint-plugin-n": "^15.7.0",
     "eslint-plugin-promise": "^6.1.1",
     "gl": "^6.0.2",
-    "jsdoc": "^3.6.11",
+    "jsdoc": "^4.0.2",
```
Comment (jordan-gillard) [inline package.json:71]: This fixes a critical bug with jsdoc version `3.6.11`. I fixed this with `npm audit fix --force`.

Comment (linev) [reply package.json:71]: I already tried version 4 of jsdoc, but it fails with template I using for jsroot documentation generation.
Before one found a solution - one need to use version 3 :(

**Comment by linev** [comment]
I merging PR. 

Thanks for contributing to the project!

**Comment by jordan-gillard** [comment]
> May be you can try to add node 20, but only with g++12?
> 
> Is it possible reduce compiler matrix only for that platform?

Sorry for this screenshot, its just that Im on mobile:
image

Compilation using all versions of g++ fails with Node 20 on Ubuntu. This may have to do with Ubuntu/Debian OS'. 

I can expand the test matrix, sure! I'll look into joining the Cern forum for ROOT and we can talk there. 

I like your suggestion of adding CI to the integration test repo you linked in the issue. Maybe that can be one of the next steps. Sound good?

**Comment by jordan-gillard** [comment]
> I mean not test all g++, but exclude g++13 on latest Ubuntu/node20.

All of the compilation stages ran separately. So I was able to test g++13 on latest Ubuntu with Node 20. Can you tell me the expected behavior?

> Just to try only g++12 on this platform. Older g++ are fully irrelevant.

Unfortunately that fails with Node 20 but passes with Node 18. Those are these jobs:
image

> But complete set of tests from https://github.com/linev/jsroot-test will be too much for CI.

How come?

**Comment by jordan-gillard** [comment]
> About node20 and Ubuntu - most probably some important packages are missing there. 
> 
> 
> 
> Looking here: https://www.npmjs.com/package/gl#system-dependencies
> 
> 
> 
> Maybe some packages listed there are missing? How one can specify them for CI?

I can edit the CI to check for & install the packages listed on the `gl` page you linked - however, if the problem was missing system dependencies, wouldn't we expect to see `gl` failing with Node 18 too? Both Node 18 & Node 20 use the same latest Ubuntu OS and same node package versions in CI. So they both use Node's builder, `node-gyp`, to build `gl`.

I believe it is more likely the problem is with `node-gyp` when using Node 20 on Debian machines.

I can check it out later this week. Maybe we should define some goals as to what we'd like CI to do!

**Comment by linev** [comment]
I am not sure, but `gl` can have different set of libs/deps for different node versions.

I using node20 and latest OpenSUSE - and it works fine for me

**Comment by jordan-gillard** [comment]
@linev Could be that OpenSUSE is RPM-based while Ubuntu is Debian 🙃

**Comment by linev** [comment]
Yes - OpenSUSE is rpm based

**Comment by linev** [comment]
@jordan-gillard 

After some gymnastics and fixing [gl](https://www.npmjs.com/package/gl) package one can test JSROOT on all platforms.

Thanks again for contribution!

**Pull request merged**